### ISSUE TITLE
[Tier 5][AMD][kernel] flash_attn_3 backward port to MI355X (gfx950) via TileLang — close GRPO actor-update gap

### ISSUE THREAD

**Issue by sunxxuns** [amdpilot-org/miles#46]
## Goal

Port the **flash_attn_3 backward** kernel to MI355X (gfx950) via
TileLang so miles GRPO training can use a real FA3-backward instead
of falling back to PyTorch SDPA. Without FA3-backward, multi-step
gradient computation through long-context attention forces a slow
SDPA path that bottlenecks the actor-update step of every GRPO
iteration.

Goal is bit-wise gradient parity vs the CUDA H200 reference at small
shapes AND measurable step-time improvement at the production shape.

## Hardware / workload

```
Hardware:    8 × AMD MI355X (single node)
Model:       amd/Llama-3.3-70B-Instruct-FP8-KV (cached) for the actor;
             tiny synthetic for the unit test
Workload:    GRPO actor update, B=2, S=8192, H=64, D=128, BF16 grads
Baseline:    current SDPA backward path (sdpa_attention.py)
Target:      backward_ms strictly lower than baseline at same gradient
             accuracy (max_abs_grad_diff < 1e-3 vs SDPA reference)
```

## Why MI355X

FA3 is Hopper-native (CUDA). The CUTE-based forward kernel was ported
in flash-attention#22 (cycle in our fork). The BACKWARD has no AMD
implementation yet — only Triton via aiter does the forward fast-path.
This issue closes the backward gap, MI355X-only because we lean on
gfx950 XDLOPs in the BWD GEMMs that MI300X can't match without a
separate tuning round.

## Scope

1. Port `flash_attn_3_bwd` from `flash-attention/hopper/` to TileLang
   targeting gfx950. Three GEMMs to fuse: `dV = P^T · dO`,
   `dP = dO · V^T`, `dS = P * (dP - rowsum)`. Mind the bf16 ULP floor
   (one bf16 ULP ≈ 0.0039) — verify max_abs_grad_diff stays under 2*ULP.
2. Add a Triton fallback that runs the same algorithm so the import
   path doesn't break on machines without TileLang.
3. Wire into `miles/python/miles/actor/attention.py` behind
   `MILES_USE_FA3_BWD=1` (auto-on for gfx950 once stable).
4. Unit test: synthetic B=1 S=512 H=16 D=64 — compute dQ, dK, dV
   three ways (SDPA, the new TileLang BWD, a hand-rolled torch
   reference) — max_abs_diff < 1e-3 across all pairs.
5. Integration test: one GRPO step on the cached Llama-3.3-70B-FP8-KV
   actor — measure `actor_update_ms` for SDPA baseline vs new path.
6. Analytical: % of the forward kernel's TFLOP/s achieved by the new
   BWD (real bound is ~50% on memory bw for the bwd; report the gap).

## Reproduction

```bash
cd /workspace/miles
# baseline
MILES_USE_FA3_BWD=0 python3 /workspace/test_harness.py
# new path
MILES_USE_FA3_BWD=1 python3 /workspace/test_harness.py
```

## Acceptance target

- `actor_update_ms` strictly lower than SDPA baseline
- `max_abs_grad_diff` < 1e-3 across {dQ, dK, dV} vs torch reference
- Unit test + integration test green
- Triton fallback runs without TileLang installed
- Doc: BWD TFLOP/s achieved + memory-bound fraction

## Notes

- Cached: `amd/Llama-3.3-70B-Instruct-FP8-KV`. Do NOT download other
  Llama variants.
- TileLang is at `/sgl-workspace/tilelang` on n08-09; verify the AMD
  fork (TileLang/tilelang main may not have gfx950 codegen yet —
  patch in-place rather than upstreaming for this trial).
- This is *framework + kernel* — touches a new TileLang kernel,
  the miles attention dispatch, a Triton fallback, and a benchmark.

**Comment by sunxxuns** [comment]
Closing — overlaps with **amdpilot-org/flash-attention#18** ("Port FA-3 backward to MI355X — single head_dim"), which is the canonical issue for the kernel port. flash-attention PR #46 is an open amdpilot-bot attempt at the same kernel (39 files, not yet merged).

The miles-side consumer wiring (this issue's framing) is genuinely separate, but it's premature to file until the upstream kernel lands. **Will refile as "Wire FA-3 backward into miles GRPO actor" after flash-attention#18 closes** — at that point the scope shrinks to ~50 lines of dispatch code + integration test, not the kernel port itself.

Apologies for the duplicate spec — caught on review. The other two new issues (sglang#25 MXFP4 prefill, sglang#26 MORI SDMA TP overlap) are clean / non-overlapping.

**Comment by jhinpan** [comment]
**AMDPilot — accumulated progress on this issue**

_Auto-generated from successive trials. Latest update: 2026-05-20T23:08:02Z by `issue-miles-46`._

### Verified infra fixes (apply during Stage 0)

_(none yet — when an agent fixes a Dockerfile / env / module-import issue, it will land here.)_


### Working approaches and their best metric

_(none yet — when a trial measurably improves the metric, the change lands here.)_


### Approaches that did NOT help — don't retry these

- Job terminated immediately due to `pending_timeout` while waiting for worker claim


### Open ideas to try next

- Investigate worker claim queue depth to diagnose resource contention
- Increase `pending_timeout` threshold in the scheduler
- Surface worker queue depth or estimated wait time in initial status updates


_amdpilot bot. Edit between the markers to add/correct fixes; the next update will preserve your changes. To disable updates on this issue, replace the body with the marker line only._
